pallets/gear-voucher/src/internal.rs

```diff
@@ -0,0 +1,202 @@
+// This file is part of Gear.
+
+// Copyright (C) 2021-2023 Gear Technologies Inc.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+#![allow(unused)]
+
+use crate::*;
+use common::{
+    storage::{Counter, CounterImpl, Mailbox},
+    DelegateFee, Origin,
+};
+use gear_core::{declare_id, ids};
+use sp_std::collections::btree_set::BTreeSet;
+
+declare_id!(VoucherId: "Voucher identifier");
+
+impl VoucherId {
+    pub fn generate<T: Config>() -> Self {
+        const SALT: &[u8] = b"voucher";
+
+        let nonce = CounterImpl::<u64, IssuedWrap<T>>::inc_get();
+
+        let argument = [SALT, &nonce.to_le_bytes()].concat();
+        ids::hash(&argument).into()
+    }
+}
+
+impl Origin for VoucherId {
+    fn into_origin(self) -> H256 {
+        self.0.into()
+    }
+
+    fn from_origin(val: H256) -> Self {
+        Self(val.to_fixed_bytes())
+    }
+}
+
+/// Type containing all data about voucher.
+#[derive(Debug, Encode, Decode, TypeInfo)]
+pub struct VoucherInfo<AccountId, BlockNumber> {
+    /// Owner of the voucher.
+    /// May be different to original issuer.
+    /// Owner manages and claims back remaining balance of the voucher.
+    pub owner: AccountId,
+    /// Set of programs this voucher could be used to interact with.
+    /// In case of None means any gear program.
```

Comment:
What about this?

```rs
    /// In case of [`None`] means any gear program.
```

See [rustdoc documentation](https://doc.rust-lang.org/rustdoc/write-documentation/linking-to-items-by-name.html#linking-to-items-by-name)